runtime/compiler/spotbugs.vim

```diff
@@ -0,0 +1,102 @@
+" Vim compiler file
+" Compiler:     Spotbugs (Java static checker; needs javac compiled classes)
+" Maintainer:   @Konfekt
+" Last Change:  2024 nov 6
+
+if exists("current_compiler") | finish | endif
+let current_compiler = "spotbugs"
+
+let s:cpo_save = &cpo
+set cpo&vim
+
+" Unfortunately Spotbugs does not output absolute paths, so you need to
+" pass the directory of the files being checked as `-sourcepath` parameter.
+" The regex, auxpath and glob try to include all dependent classes of the
+" current buffer. See https://github.com/spotbugs/spotbugs/issues/856
+
+" FIXME: When "search()" is used with the "e" flag, it makes no _further_
+" progress after claiming an EOL match (i.e. "\_" or "\n", but not "$").
+" XXX: Omit anonymous class declarations
+let s:types = '\<\%(\.\@1<!class\|@\=interface\|enum\|record\)\%(\s\|$\)'
+let s:names = '\<\%(\.\@1<!class\|@\=interface\|enum\|record\)\s*\(\K\k*\)\>'
+let s:slash = exists('+shellslash') && !&shellslash ? '\' : '/'
+
+if has('syntax') && exists('g:syntax_on') && exists('b:current_syntax') &&
+    \ b:current_syntax == 'java' && hlexists('javaClassDecl')
+
+  function! s:GetDeclaredTypeNames() abort
+    defer execute('keepjumps normal ``')
+    normal gg
+    let type_names = []
+    let lnum = search(s:types, 'eW')
+    while lnum > 0
+      if synIDattr(synID(lnum, (col('.') - 1), 0), 'name') == 'javaClassDecl'
+        let tokens = matchlist(getline(lnum)..getline(lnum + 1), s:names)
+        if !empty(tokens) | call add(type_names, tokens[1]) | endif
+      endif
+      let lnum = search(s:types, 'eW')
+    endwhile
+    return type_names
+  endfunction
+
+else
+
+  function! s:GetDeclaredTypeNames() abort
+    " Undo the unsetting of &hls, see below
+    if &hls | defer execute('set hls') | endif
+    " Copy buffer contents for modification
+    silent %y
+    new
+    defer execute('silent bwipeout')
+    " Apply ":help scratch-buffer" effects
+    setlocal buftype=nofile bufhidden=hide noswapfile nohls
+    silent normal P
+    let @" = ""
```

Comment:
@zzzyxwvut Does this preserve the register contents?

Reply:
These register assignments ought to discard the yanked and  
the put whole buffer contents.

https://github.com/vim/vim/blob/7724d621f21ff6599ba49f11401ddb45ff321099/runtime/doc/change.txt#L1273-L1274

https://github.com/vim/vim/blob/7724d621f21ff6599ba49f11401ddb45ff321099/runtime/doc/change.txt#L1264-L1266

Reply:
Yes, discarding it done, but what about restoring those previously there? Wouldn't the user be surprised to find that changing the compiler clears the paste register?

Reply:
Especially so since `:compiler spotbugs` demands continual resourcing

Reply:
Let's then use a named register and manage its contents.

```diff
diff --git a/runtime/compiler/spotbugs.vim b/runtime/compiler/spotbugs.vim
index 137c98cf2..6d4e727a7 100644
--- a/runtime/compiler/spotbugs.vim
+++ b/runtime/compiler/spotbugs.vim
@@ -44,15 +44,15 @@ else
   function! s:GetDeclaredTypeNames() abort
     " Undo the unsetting of &hls, see below
     if &hls | defer execute('set hls') | endif
+    " Possibly restore the current value for register "y", see below
+    defer execute('let @y = '..(!empty(string(@y)) ? string(@y) : '""'))
     " Copy buffer contents for modification
-    silent %y
+    silent %y y
     new
     defer execute('silent bwipeout')
     " Apply ":help scratch-buffer" effects
     setlocal buftype=nofile bufhidden=hide noswapfile nohls
-    silent normal P
-    let @" = ""
-    let @0 = ""
+    silent normal "yP
     " Discard text blocks and strings
     silent keeppatterns %s/\\\@<!"""\_.\{-}\\\@<!"""\|\\"//ge
     silent keeppatterns %s/".*"//ge
```

Reply:
Thank you!